File: app/src/main/java/com/muratcan/apps/petvaccinetracker/adapter/PetAdapter.java
package com.muratcan.apps.petvaccinetracker.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.muratcan.apps.petvaccinetracker.R;
import com.muratcan.apps.petvaccinetracker.model.Pet;
import com.muratcan.apps.petvaccinetracker.util.ImageUtils;

import java.util.ArrayList;
import java.util.List;

public class PetAdapter extends RecyclerView.Adapter<PetAdapter.PetViewHolder> {
    private final List<Pet> pets;
    private final OnPetClickListener listener;

    public interface OnPetClickListener {
        void onPetClick(Pet pet);
    }

    public PetAdapter(List<Pet> pets, OnPetClickListener listener) {
        this.pets = new ArrayList<>(pets);
        this.listener = listener;
    }

    @NonNull
    @Override
    public PetViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_pet, parent, false);
        return new PetViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull PetViewHolder holder, int position) {
        holder.bind(pets.get(position), listener);
    }

    @Override
    public int getItemCount() {
        return pets.size();
    }

    public void updatePets(List<Pet> newPets) {
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(new PetDiffCallback(pets, newPets));
        pets.clear();
        pets.addAll(newPets);
        diffResult.dispatchUpdatesTo(this);
    }

    private static class PetDiffCallback extends DiffUtil.Callback {
        private final List<Pet> oldPets;
        private final List<Pet> newPets;

        PetDiffCallback(List<Pet> oldPets, List<Pet> newPets) {
            this.oldPets = oldPets;
            this.newPets = newPets;
        }

        @Override
        public int getOldListSize() {
            return oldPets.size();
        }

        @Override
        public int getNewListSize() {
            return newPets.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldPets.get(oldItemPosition).getId() == newPets.get(newItemPosition).getId();
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            Pet oldPet = oldPets.get(oldItemPosition);
            Pet newPet = newPets.get(newItemPosition);
            return oldPet.getName().equals(newPet.getName()) &&
                   oldPet.getType().equals(newPet.getType()) &&
                   oldPet.getBreed().equals(newPet.getBreed()) &&
                   (oldPet.getImageUri() == null ? 
                        newPet.getImageUri() == null : 
                        oldPet.getImageUri().equals(newPet.getImageUri()));
        }
    }

    public static class PetViewHolder extends RecyclerView.ViewHolder {
        private final ImageView petImageView;
        private final TextView petNameTextView;
        private final TextView petTypeTextView;
        private final TextView petBreedTextView;
        private final View cardView;

        public PetViewHolder(@NonNull View itemView) {
            super(itemView);
            cardView = itemView.findViewById(R.id.cardView);
            petImageView = itemView.findViewById(R.id.petImageView);
            petNameTextView = itemView.findViewById(R.id.petNameTextView);
            petTypeTextView = itemView.findViewById(R.id.petTypeTextView);
            petBreedTextView = itemView.findViewById(R.id.petBreedTextView);
        }

        void bind(Pet pet, OnPetClickListener listener) {
            String transitionName = "pet_card_" + pet.getId();
            cardView.setTransitionName(transitionName);
            cardView.setTag(transitionName);

            petNameTextView.setText(pet.getName());
            petTypeTextView.setText(pet.getType());
            petBreedTextView.setText(pet.getBreed());

            if (pet.getImageUri() != null) {
                ImageUtils.loadImage(itemView.getContext(), pet.getImageUri(), petImageView);
            } else {
                petImageView.setImageResource(R.drawable.ic_pet_placeholder);
            }

            itemView.setOnClickListener(v -> listener.onPetClick(pet));
        }
    }
} 

File: app/src/main/java/com/muratcan/apps/petvaccinetracker/adapter/PetAdapter.kt
package com.muratcan.apps.petvaccinetracker.adapter

import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.chip.Chip
import com.muratcan.apps.petvaccinetracker.R
import com.muratcan.apps.petvaccinetracker.model.Pet
import com.muratcan.apps.petvaccinetracker.util.ImageUtils

class PetAdapter(
    pets: List<Pet>,
    private val listener: OnPetClickListener
) : RecyclerView.Adapter<PetAdapter.PetViewHolder>() {

    private val pets: MutableList<Pet> = ArrayList(pets)

    fun interface OnPetClickListener {
        fun onPetClick(pet: Pet)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PetViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_pet, parent, false)
        return PetViewHolder(view)
    }

    override fun onBindViewHolder(holder: PetViewHolder, position: Int) {
        holder.bind(pets[position], listener)
    }

    override fun getItemCount(): Int {
        return pets.size
    }

    fun updatePets(newPets: List<Pet>) {
        val diffResult = DiffUtil.calculateDiff(PetDiffCallback(pets, newPets))
        pets.clear()
        pets.addAll(newPets)
        diffResult.dispatchUpdatesTo(this)
    }

    private class PetDiffCallback(
        private val oldPets: List<Pet>,
        private val newPets: List<Pet>
    ) : DiffUtil.Callback() {

        override fun getOldListSize(): Int {
            return oldPets.size
        }

        override fun getNewListSize(): Int {
            return newPets.size
        }

        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            return oldPets[oldItemPosition].id == newPets[newItemPosition].id
        }

        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
            val oldPet = oldPets[oldItemPosition]
            val newPet = newPets[newItemPosition]
            return oldPet.name == newPet.name &&
                    oldPet.type == newPet.type &&
                    oldPet.breed == newPet.breed &&
                    oldPet.imageUri == newPet.imageUri
        }
    }

    class PetViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val cardView: View = itemView.findViewById(R.id.cardView)
        private val petImageView: ImageView = itemView.findViewById(R.id.petImageView)
        private val petNameTextView: TextView = itemView.findViewById(R.id.petNameTextView)
        private val petTypeChip: Chip = itemView.findViewById(R.id.petTypeChip)
        private val petBreedTextView: TextView = itemView.findViewById(R.id.petBreedTextView)

        fun bind(pet: Pet, listener: OnPetClickListener) {
            val transitionName = "pet_card_${pet.id}"
            cardView.transitionName = transitionName
            cardView.tag = transitionName

            petNameTextView.text = pet.name
            petTypeChip.text = pet.type
            petBreedTextView.text = pet.breed

            if (pet.imageUri != null) {
                ImageUtils.loadImage(itemView.context, Uri.parse(pet.imageUri), petImageView)
            } else {
                petImageView.setImageResource(R.drawable.ic_pet_placeholder)
            }

            itemView.setOnClickListener { listener.onPetClick(pet) }
        }
    }
} 